// azure-devops-ui/Components/Panel/CustomPanel.js
import * as tslib_1 from "tslib";
import "../../CommonImports";
import "../../Core/core.css";
import "./Panel.css";
import * as React from "react";
import { ScreenContext } from '../../Core/Util/Screen';
import { Callout } from '../../Callout';
import { Observer } from '../../Observer';
import { Surface, SurfaceContext } from '../../Surface';
import { css } from '../../Util';
var CustomPanel = /** @class */ (function (_super) {
    tslib_1.__extends(CustomPanel, _super);
    function CustomPanel(props) {
        var _this = _super.call(this, props) || this;
        _this.calloutContentRef = React.createRef();
        _this.defaultActiveElement = function () {
            // We don't ever want the Panel to set focus to the body, so if the defaultActiveElement
            // prop that is provided cannot be found, instead use the panel's default focus element.
            var defaultActiveElement = _this.props.defaultActiveElement;
            var selector = typeof defaultActiveElement === "function" ? defaultActiveElement() : defaultActiveElement;
            if (selector && _this.calloutContentRef.current) {
                var matches = _this.calloutContentRef.current.querySelectorAll(selector);
                if (matches && matches.length) {
                    return selector;
                }
            }
            return ".bolt-panel-focus-element";
        };
        // TODO: Change to false in M150
        _this.state = {
            isDisplayed: true
        };
        return _this;
    }
    CustomPanel.prototype.render = function () {
        var _this = this;
        var _a = this.props, ariaLabel = _a.ariaLabel, ariaLabelledBy = _a.ariaLabelledBy, blurDismiss = _a.blurDismiss, calloutClassName = _a.calloutClassName, children = _a.children, className = _a.className, contentClassName = _a.contentClassName, escDismiss = _a.escDismiss, id = _a.id, lightDismiss = _a.lightDismiss, modal = _a.modal, onDismiss = _a.onDismiss, _b = _a.size, size = _b === void 0 ? 1 /* Medium */ : _b;
        var isDisplayed = this.state.isDisplayed;
        return (React.createElement(Observer, { size: this.context.size }, function (props) {
            var fullscreen = props.size === 0 /* xsmall */;
            return (React.createElement(Callout, { ariaLabel: ariaLabel, ariaLabelledBy: ariaLabelledBy, blurDismiss: blurDismiss, className: css("bolt-panel", calloutClassName, isDisplayed === false && "no-events"), contentClassName: css(contentClassName, "bolt-panel-callout-content scroll-auto", isDisplayed ? "in" : "out", fullscreen ? "bolt-panel-fullscreen absolute-fill" : "relative"), contentJustification: 3 /* Stretch */, contentLocation: 2 /* End */, contentOrientation: 0 /* Column */, contentRef: _this.calloutContentRef, contentShadow: true, contentSize: fullscreen ? undefined : size, escDismiss: escDismiss, id: id, focuszoneProps: {
                    circularNavigation: true,
                    defaultActiveElement: _this.defaultActiveElement,
                    focusOnMount: true,
                    handleTabKey: true,
                    includeDefaults: true,
                    postprocessKeyStroke: function (event) {
                        // We want to prevent moving outside the panel if there are no focusable elements in the panel.
                        event.which === 9 /* tab */ && event.preventDefault();
                        return 2 /* IgnoreParents */;
                    }
                }, lightDismiss: lightDismiss && isDisplayed !== false, modal: modal && isDisplayed !== false, onDismiss: onDismiss },
                React.createElement(SurfaceContext.Consumer, null, function (surfaceContext) { return (React.createElement(Surface, tslib_1.__assign({}, surfaceContext, { spacing: 1 /* default */ }),
                    React.createElement("div", { className: css(className, "bolt-panel-root flex-column flex-grow scroll-hidden") },
                        React.createElement("div", { className: "bolt-panel-focus-element no-outline", tabIndex: -1 }),
                        children))); })));
        }));
    };
    // TODO: Uncomment in M150
    /* public componentDidMount() {
        setTimeout(() => {
            this.setState({ isDisplayed: true });
        }, 0);
    }

    public componentWillUnmount() {
        if($DEBUG) {
            if(this.state.isDisplayed === true) {
                console.error("Panel was unmounted while still displayed; call animateOut first and unmount after the returned promise resolves");
            }
        }
    } */
    CustomPanel.prototype.animateOut = function () {
        return Promise.resolve();
        // TODO: Uncomment in M150
        /* return new Promise(resolve => {
            this.setState({ isDisplayed: false });

            setTimeout(() => {
                resolve();
            }, 200);
        }); */
    };
    CustomPanel.defaultProps = {
        escDismiss: true,
        lightDismiss: true,
        modal: true
    };
    CustomPanel.contextType = ScreenContext;
    return CustomPanel;
}(React.Component));
export { CustomPanel };
