// azure-devops-ui/Components/Table/ColumnMore.js
import "../../CommonImports";
import "../../Core/core.css";
import "./Table.css";
import * as React from "react";
import { ObservableValue } from '../../Core/Observable';
import { ExpandableButton } from '../../Button';
import { ContextualMenu } from '../../Menu';
import * as Resources from '../../Resources.Widgets';
import { css, eventTargetContainsNode, preventDefault } from '../../Util';
/**
 * The ColumnMore class is used as a custom column within a Table.
 *
 * It renders a fixed width column with a button that is visible on focus and
 * hover that will show a contextual menu for the given item. The caller
 * should create a new instance of this for each table. The header for this
 * column is empty.
 *
 * When creating the column you must supply delegate that is used to get the
 * menuProps when the menu is opened.
 */
var ColumnMore = /** @class */ (function () {
    /**
     * Creates an instance of ColumnMore with the appropriate delegates.
     *
     * @param menuProvider Method that returns the menu props. This is called
     *  with the item from the row the menu was activated on.
     * @param menuAvailable Optional method that is called when the row
     *  is rendered. The MoreButton will be shown if no method is supplied
     *  or if the method is supplied and it returns true.
     */
    function ColumnMore(menuProvider, menuAvailable, onActiviate) {
        var _this = this;
        this.ariaLabel = Resources.MoreActions;
        this.columnLayout = 0 /* none */;
        this.id = "_more";
        this.width = new ObservableValue(2.625);
        this.widthStyle = 1 /* REM */;
        this.renderCell = function (rowIndex, columnIndex, tableColumn, listItem) {
            return (React.createElement("td", { "aria-colindex": columnIndex + 1, className: css("bolt-table-cell-side-action bolt-table-cell bolt-list-cell", "col-" + columnIndex), "data-column-index": columnIndex, key: "col-more", onClick: _this.onClick, onDoubleClick: preventDefault, onKeyDown: _this.onKeyDown, onMouseDown: preventDefault },
                React.createElement("div", { className: "bolt-table-cell-content-reveal flex-row justify-center" }, (!_this.menuAvailable || _this.menuAvailable(listItem)) && (React.createElement(ExpandableButton, { ariaLabel: Resources.MoreTooltip, className: "bolt-table-button-more", excludeTabStop: true, hideDropdownIcon: true, iconProps: { className: "small", iconName: "MoreVertical" }, onClick: function (e) {
                        if (_this.onActivate) {
                            _this.onActivate(rowIndex, columnIndex);
                        }
                    }, renderCallout: function (dropdown, dropdownId, anchorElement) {
                        return (React.createElement(ContextualMenu, { anchorElement: anchorElement, anchorOrigin: { horizontal: "end" /* end */, vertical: "end" /* end */ }, menuProps: _this.menuProvider(listItem), menuOrigin: { horizontal: "end" /* end */, vertical: "start" /* start */ }, onDismiss: dropdown.collapse }));
                    }, subtle: true, tooltipProps: { text: Resources.MoreTooltip } })))));
        };
        this.renderHeaderCell = function (columnIndex, tableColumn) {
            return (React.createElement("th", { "aria-colindex": columnIndex + 1, "aria-label": tableColumn.ariaLabel, className: css(tableColumn.headerClassName, "bolt-table-header-cell bolt-table-header-cell-empty", "col-header-" + columnIndex), "data-column-index": columnIndex, key: "col-more" }));
        };
        this.onClick = function (event) {
            if (!event.defaultPrevented) {
                if (!eventTargetContainsNode(event, ["A"], document.body)) {
                    event.preventDefault();
                }
            }
        };
        this.onKeyDown = function (event) {
            if (!event.defaultPrevented) {
                if (event.which === 13 /* enter */ || event.which === 32 /* space */) {
                    if (!eventTargetContainsNode(event, ["A"])) {
                        event.preventDefault();
                    }
                }
            }
        };
        this.menuAvailable = menuAvailable;
        this.menuProvider = menuProvider;
        this.onActivate = onActiviate;
    }
    return ColumnMore;
}());
export { ColumnMore };
